packages/did/src/DidLinks/AccountLinks.chain.ts

```diff
@@ -194,5 +192,66 @@ export async function associateAccountToChainArgs(
       'Ethereum linking is not yet supported by this chain'
     )
 
-  return [accountAddress, validTill, proof]
+  return [accountAddress, validUntil, proof]
+}
+
+/**
+ * Wraps the provided challenge according to the key type.
+ *
+ * Ethereum addresses will cause the challenge to be prefixed with
+ * `\x19Ethereum Signed Message:\n` and the length of the message.
+ *
+ * For all other key types the message will be wrapped in `<Bytes>..</Bytes>`.
+ *
+ * @param type The key type that will sign the challenge.
+ * @param challenge The challenge to proof ownership of both account and DID.
+ * @returns The wrapped challenge.
+ */
+export function getWrappedChallenge(
+  type: KeypairType,
+  challenge: Uint8Array
+): Uint8Array {
+  if (type === 'ethereum') {
+    const length = stringToU8a(String(challenge.length))
+    return u8aConcatStrict([U8A_WRAP_ETHEREUM, length, challenge])
+  }
+  return u8aWrapBytes(challenge)
+}
+
+/**
+ * Builds the parameters for an extrinsic to link the `account` to the `did` where the fees and deposit are covered by some third account.
+ * This extrinsic must be authorized using the same full DID.
+ * Note that in addition to the signing account and DID used here, the submitting account will also be able to dissolve the link via reclaiming its deposit!
+ *
+ * @param accountAddress Address of the account to be linked.
+ * @param did Full DID to be linked.
+ * @param sign The sign callback that generates the account signature over the encoded (DidAddress, BlockNumber) tuple.
+ * @param nBlocksValid The link request will be rejected if submitted later than (current block number + nBlocksValid)?
+ * @returns An array of parameters for `api.tx.didLookup.associateAccount` that must be DID-authorized by the full DID used.
+ */
+export async function associateAccountToChainArgs(
+  accountAddress: Address,
+  did: DidUri,
+  sign: (encodedLinkingDetails: HexString) => Promise<Uint8Array>,
+  nBlocksValid = 10
+): Promise<AssociateAccountToChainResult> {
+  const api = ConfigService.get('api')
+
+  const blockNo = await api.query.system.number()
+  const validTill = blockNo.addn(nBlocksValid)
+
+  const challenge = await getLinkingChallenge(did, validTill)
+
+  const predictedType = accountAddress.length === 20 ? 'ethereum' : 'sr25519'
```

Comment:
wait we can't just assume it's an sr25519 address can we?

Reply:
There are only two different wrapping strategies. `ethereum` and the substrate way (`sr25519`, `ed25519`, `ecdsa`). So for anything not ethereum, we can just pretend it's one of the others. Doesn't make any difference which one we choose.

But yeah it's an ugly place in code. Don't really like it as well.